pkg/conversion/deep_copy.go

```diff
@@ -17,52 +17,104 @@ limitations under the License.
 package conversion
 
 import (
-	"bytes"
-	"encoding/gob"
+	"fmt"
 	"reflect"
-	"sync"
 )
 
-// pool is a pool of copiers
-var pool = sync.Pool{
-	New: func() interface{} { return newGobCopier() },
-}
-
 // DeepCopy makes a deep copy of source or returns an error.
 func DeepCopy(source interface{}) (interface{}, error) {
-	v := reflect.New(reflect.TypeOf(source))
+	v, err := deepCopy(reflect.ValueOf(source))
+	return v.Interface(), err
+}
 
-	c := pool.Get().(gobCopier)
-	defer pool.Put(c)
-	if err := c.CopyInto(v.Interface(), source); err != nil {
-		return nil, err
-	}
+func deepCopy(src reflect.Value) (reflect.Value, error) {
+	switch src.Kind() {
+	case reflect.Chan, reflect.Func, reflect.UnsafePointer, reflect.Uintptr:
+		return src, fmt.Errorf("cannot deep copy kind: %s", src.Kind())
 
-	return v.Elem().Interface(), nil
-}
+	case reflect.Array:
+		dst := reflect.New(src.Type())
+		for i := 0; i < src.Len(); i++ {
+			copyVal, err := deepCopy(src.Index(i))
+			if err != nil {
+				return src, err
+			}
+			dst.Elem().Index(i).Set(copyVal)
+		}
+		return dst.Elem(), nil
 
-// gobCopier provides a copy mechanism for objects using Gob.
-// This object is not safe for multiple threads because buffer
-// is shared.
-type gobCopier struct {
-	enc *gob.Encoder
-	dec *gob.Decoder
-}
+	case reflect.Interface:
+		if src.IsNil() {
+			return src, nil
+		}
+		dst := reflect.Zero(src.Type())
+		copyVal, err := deepCopy(src.Elem())
+		if err != nil {
+			return src, err
+		}
+		dst.Set(copyVal)
+		return dst, nil
 
-func newGobCopier() gobCopier {
-	buf := &bytes.Buffer{}
-	return gobCopier{
-		enc: gob.NewEncoder(buf),
-		dec: gob.NewDecoder(buf),
-	}
-}
+	case reflect.Map:
+		if src.IsNil() {
+			return src, nil
+		}
+		dst := reflect.MakeMap(src.Type())
+		for _, k := range src.MapKeys() {
+			copyVal, err := deepCopy(src.MapIndex(k))
+			if err != nil {
+				return src, err
+			}
+			dst.SetMapIndex(k, copyVal)
```

Comment:
This doesn't "deep-copy" keys of the map. It works for all our current api types, but this might not be true in the future, right? Maybe a TODO to fix it?